### PR TITLE
#272 - Dynamic Form Template

### LINKED ISSUE #272: New/edit run form support
### Description

This first and most important part of this ticket is creating a form template that we will be able to re-use. 

Then this generic form should be used via a pop up when the user clicks on the new button labeled "edit current run". 
See Proposed Solution for more info.

The form should be available via a pop up / dialogue. 
See Proposed Solution for more info.

We will define and auto fill the Run content via an endpoint that is already hooked up to the backend. 
See Proposed Solution for more info.

### Acceptance Criteria

Just get something up with a generic form template component, and I can let you know if there's anything we could incorporate from Angular 18.

Ultimately this PR should add a generic form template, and add support via that form template for editing and creating new runs with driver, location, and notes for a run. 

### Proposed Solution

I recommend to complete in order of:
1. Creating the form (use mock data to get started, just make a static run variable)
2. Creating the button (this should only require putting the form inside a modal)
3. Hooking up the data. (should be the last step and simple replace the mock data, and be re-queried each time the user opens the form)

--

1. ##### Creating the Form
Here is essentially what we need just make sure to convert to Angular 18 convention (@for, and @if instead of ngFor and ngIf): https://github.com/Sharmil001/angular-dynamic-components/blob/master/src/app/dynamic-form/dynamic-form.component.ts
here is the article for it: https://medium.com/@sharmiladroja17/angular-dynamic-forms-dynamic-forms-component-b4cfadbdccf0

2. ##### Creating the button 

See the existing new run button for css styling, and basic button structure. This button should be it's own component that controls a modal pop up that houses the form for creating a new run.

Here is a video showing different options for creating modals in angular,
I recommend using signals, that approach to pop ups is shown at 14 minutes (no need to use the bootstrap he is using but the basic setup of the styles and signals is useful): https://youtu.be/7BK4Wgc6w0w?t=840

here is some sample code for something like the pop up, to be used when the user clicks on the new run button: 
html:
```
<!-- Modal Overlay -->
  @if (systemConfigOpen()) {
    <div class="modal-overlay" (click)="closeSystemConfig()">
      <!-- the even stopProgagation just is there to close the modal when the user clicks outside of it  -->
      <div class="modal-content" (click)="$event.stopPropagation()">
        <!-- FORM COMPONENT GOES HERE -->
        <system-config [onSubmitCallback]="onSubmitCallback"></system-config>
        <!-- closes the form -->
        <button (click)="closeSystemConfig()">Close</button>
      </div>
    </div>
```

ts:
```
export class SmsControlCenterComponent {
  // Signal to track whether the system-config modal is open
  systemConfigOpen = signal(false);

  openSystemConfig = () => {
    this.systemConfigOpen.set(true);
  }

  closeSystemConfig = () => {
    this.systemConfigOpen.set(false);
  }

    // Define the callback as an arrow function to preserve 'this'
  onSubmitCallback = async () => {
    console.log('trying to close');
    this.closeSystemConfig();
  }
}
```

3. ##### Where to find Run data for the form 
new run values are all available via an already defined via (getRunById() or getLatetestRun(), see usage via key word search). 
From those functions you get this data (found in types.utils.ts): 
  ```
  export type Run = {
    id: number;
    locationName: string;
    driverName: string;
    time: Date;
    notes: string;
  };
  ```

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by justinsoto**
## Changes

Added "Edit Current Run" button to the nav bar.
Clicking the button opens a dialog for the template form.
Each input field has an independent validator.
Submitting the form prints all the input data to the console.

## Test Cases

- Case A
- Edge case
- ...

## Screenshots

Argos Form Template

## To Do
- Create an enum file for different fields, and lists of fields for future use.
- Find a way to emit the form to the caller (openDialog caller).

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `package-lock.json` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes # (issue #)

### CODE REVIEW

File: angular-client/src/app/app-nav-bar/app-nav-bar.component.ts [hunk 1]
```diff
@@ -58,6 +62,41 @@ export class AppNavBarComponent implements OnInit {
     });
   };
 
+  renderTemplate = () => {
+    // console.log("edit run clicked.");
+    this.dialogService.open(RunFormTemplateComponent, {
```
Comment (bracyw) [inline angular-client/src/app/app-nav-bar/app-nav-bar.component.ts]: I like your idea of just creating a separate form data file that defines these things. Also maybe having a the dialogService open another component that houses the form fields and inputs them into the form component itself (just a wrapper around the form). To avoid overloading top bar. We can talk about it more tuesday.

File: angular-client/src/components/run-form-template/run-form-template.component.css [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+.error-msg {
+  color: red;
+  margin-top: 4px;
+  font-size: 12px;
+  font-weight: 500;
+}
+
+
+::ng-deep .dynamic-form .p-dropdown,
```
Comment (bracyw) [inline angular-client/src/components/run-form-template/run-form-template.component.css]: These I believe can be deleted all the ng-deep stuff and maybe the whole file. Check and see.

File: angular-client/src/components/run-form-template/run-form-template.component.html [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        [ngClass]="{
+          'ng-invalid ng-dirty':
+            form.get(field.name)?.invalid && (form.get(field.name)?.dirty || form.get(field.name)?.touched)
+        }"
+      > -->
+      <!-- </p-dropdown> -->
+
+      <!-- Checkbox -->
+      <!-- <p-checkbox
+        *ngIf="field.type === 'checkbox'"
```
Comment (bracyw) [inline angular-client/src/components/run-form-template/run-form-template.component.html]: This can all be deleted.

File: angular-client/src/components/run-form-template/run-form-template.component.html [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+      <div
+        class="error-msg"
+        *ngIf="form.get(field.name)?.invalid && (form.get(field.name)?.dirty || form.get(field.name)?.touched)"
+      >
+        {{ getErrorMessage(field) }}
+      </div>
+    </div>
+  }
+
+  <button pButton type="button" [disabled]="!form.valid" label="Submit" (click)="onSubmit()"></button>
```
Comment (bracyw) [inline angular-client/src/components/run-form-template/run-form-template.component.html]: Here you can replace this button with a toast-button... but we can do that when merging. (toast button has been updated recently)

File: angular-client/src/components/run-form-template/run-form-template.component.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+  public config = inject(DynamicDialogConfig);
+
+  fields: DynamicFormField[] = [
+    // Form Fields For Login
+    // {
+    //   name: 'full name',
+    //   label: 'Full Name',
+    //   type: 'text',
+    //   placeholder: 'Enter Full Name',
+    //   required: true,
```
Comment (bracyw) [inline angular-client/src/components/run-form-template/run-form-template.component.ts]: can delete

File: angular-client/src/components/run-form-template/run-form-template.component.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    //   this.buildForm();
+    // }
+  }
+
+  buildForm() {
+    const group: any = {};
+    this.fields.forEach((field) => {
+      const control = this.fb.control({
+        value: this.formData ? this.formData.get(field.name) : '',
+        disabled: field.disabled //|| !this.isEdit
```
Comment (bracyw) [inline angular-client/src/components/run-form-template/run-form-template.component.ts]: maybe just yeat this.isEdit

File: angular-client/src/pages/landing-page/landing-page.component.html [hunk 7]
```diff
@@ -9,6 +9,7 @@
         </mat-grid-tile>
         <mat-grid-tile [colspan]="1" rowspan="4">
           <connection-display style="height: 100%; width: 100%" />
+          <!-- <run-form-template (submitForm)="saveForm($event)"></run-form-template> -->
```
Comment (bracyw) [inline angular-client/src/pages/landing-page/landing-page.component.html]: delete

File: angular-client/src/pages/landing-page/landing-page.component.ts [hunk 8]
```diff
@@ -45,4 +46,9 @@ export default class LandingPageComponent implements OnInit {
   onResize() {
     this.isMobile = window.innerWidth <= this.mobileThreshold;
   }
+
+  saveForm(formData: FormGroup): void {
```
Comment (bracyw) [inline angular-client/src/pages/landing-page/landing-page.component.ts]: This is what you should now call in your parent component, to get the form and process it. (e.g. actually call the create new run endpoint). It should be moved to new abstracted component mentioned above.

**Review by bracyw — CHANGES_REQUESTED** [review]
Looking fire. We can merge Friday and might be able to add some fancy primeng input fields for styling... so don't worry styling until then.

**Review by bracyw — APPROVED** [review]
